phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ImmutableIndexExtendedIT.java

```diff
@@ -0,0 +1,258 @@
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class ImmutableIndexExtendedIT extends BaseUniqueNamesOwnClusterIT {
+
+    private final String tableDDLOptions;
+
+    public ImmutableIndexExtendedIT() {
+        StringBuilder optionBuilder = new StringBuilder("IMMUTABLE_ROWS=true");
+        this.tableDDLOptions = optionBuilder.toString();
+    }
+
+    @BeforeClass
+    public static void doSetup() throws Exception {
+        Map<String, String> props = Maps.newHashMapWithExpectedSize(1);
+        props.put(HConstants.HBASE_CLIENT_RETRIES_NUMBER, "5");
+        setUpTestDriver(new ReadOnlyProps(props));
+    }
+
+    public static class PreMutationFailingRegionObserver extends SimpleRegionObserver {
+
+        @Override public void preBatchMutate(ObserverContext<RegionCoprocessorEnvironment> c,
+                MiniBatchOperationInProgress<Mutation> miniBatchOp) throws IOException {
+            System.out.println("preBatchMutate my coproc " + this);
+            throw new IOException();
+        }
+    }
+
+    public static class PostMutationFailingRegionObserver extends SimpleRegionObserver {
+        
+        @Override public void postBatchMutate(ObserverContext<RegionCoprocessorEnvironment> c,
+                MiniBatchOperationInProgress<Mutation> miniBatchOp) throws IOException {
+            System.out.println("postBatchMutate my coproc " + this);
+            throw new IOException();
+        }
+    }
+
+    public static class FailOnceMutationRegionObserver extends SimpleRegionObserver {
+
+        private boolean failOnce = true;
+
+        @Override public void preBatchMutate(ObserverContext<RegionCoprocessorEnvironment> c,
+                MiniBatchOperationInProgress<Mutation> miniBatchOp) throws IOException {
+            System.out.println("Fail once my coproc " + this);
+            if (failOnce) {
+                // next attempt don't raise
+                failOnce = false;
+                throw new IOException();
+            }
+        }
+    }
+
+    private void createAndPopulateTable(Connection conn, String tableName, int rowCount)
+            throws Exception {
+        String ddl = "CREATE TABLE " + tableName +
+                " (id integer not null primary key, val1 varchar(10), val2 varchar(10), val3 varchar(10))"
+                + tableDDLOptions;
+
+        conn.createStatement().execute(ddl);
+
+        String dml = "UPSERT INTO " + tableName + " (id, val1, val2, val3) VALUES (?, ?, ?, ?)";
+        PreparedStatement stmt = conn.prepareStatement(dml);
+
+        for (int id = 1; id <= rowCount; ++id) {
+            stmt.setInt(1, id);
+            stmt.setString(2, "a" + id);
+            stmt.setString(3, "ab" + id);
+            stmt.setString(4, "abc" + id);
+            stmt.executeUpdate();
+        }
+        conn.commit();
+    }
+
+    private void createIndex(Connection conn, String dataTable, String indexTable)
+            throws Exception {
```

Comment:
Is there already a function like this another IT? If you can find one, please use that one